File: app/maptile/parse_maptiles.rb
#!/usr/bin/env ruby -wKU

infile = ARGV[0] || "#{Dir.pwd}/Maptiles.sql"
dir    = "#{Dir.pwd}/maptiles-output"
Dir.mkdir(dir, 0755) if not Dir.exists?(dir)

File.open(infile) { |file|
  file.readlines.each { |line| 
    next unless line =~ /^INSERT/
    fields = line.split(/\(|\)/)[1]
    zoom, x, y, flags, length, image = fields.split(/\,/)
    puts "#{zoom}, #{x}, #{y}, #{flags}, #{length}"
    # data = image.sub(/^X'/, '').sub(/'$/, '')
    data = image.sub(/^X'|'$/, '')
    next if image.length < 128
    
    filename = "#{dir}/#{x},#{y}\@#{zoom}.png"
    File.open(filename, "w") { |f2|
      0.step(image.length, 2) { |i| f2 << "0x#{data[i, 2]}".hex.chr }
    }
  }
}

File: app/maptile/parse_maptiles.pl
#!/usr/bin/env perl

use strict;
use vars qw { $FILE };

$FILE = shift;
if ($FILE eq "") {
    die "Syntax: $0 [filename]\n";
}

&parse($FILE);

sub parse {
    my ($FILE) = @_;
    open(FILE, "<$FILE") || die "$FILE: $!";
    mkdir("./maptiles-output", 0755);
    while (<FILE>) {
        chomp;
        my $j = 0;
        my $contents = $_;
        next unless($contents =~ /^INSERT /);
        my ($junk, $sql, $junk) = split(/\(|\)/, $contents);
        my ($zoom, $x, $y, $flags, $length, $data) = split(/\,/, $sql);
        $data =~ s/^X'//;
        $data =~ s/'$//;
        my $filename = "./maptiles-output/$x,$y\@$zoom.png";
        next if int(length($data)) < 128;
        print $filename . "\n";
        open(OUT, ">$filename") . "\n" || die "$filename: $!";
        print int(length($data)) . "\n";
        while($j < length($data)) {
            my $hex = "0x" . substr($data, $j, 2);
            print OUT chr(hex($hex));
            $j += 2;
        }
        close(OUT);
    }
    close(FILE);
}

